### DISCUSSION TITLE
FluentAvalonia ListBox Selection Bug

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#8634] [Q&A]
My Theme work with FluentAvalonia is going pretty well. There is just one little annoying bug.

In a bound ListBox, if I delete the last element and select the previous one, it is selected but does not appear as selected.

Is that a bug on the FluentAvalonia side or a bug on the Avalonia side? What could be causing this? FluentAvalonia relies purely on Styles so it's the styles engine that should handle it.

If necessary, I could produce a minimal reproduction project. I haven't yet tested whether the problem was dependent on the way I bind or can be reproduced out-of-the-box.

**Comment by timunie** [comment]
A minimal sample is always helpful. Also you could easily test without FluentAvalonia Styles to see if it's a bug there.

**Comment by mysteryx93** [comment]
I investigated this further and the problem isn't related to FluentAvalonia at all. Here's a sample project.

I have a CollectionView that manages the list and selection in a way similar to WPF.

The problem is that I'm listening to Source_CollectionChanged to set the new selection; then there's a built-in ICollectionChangedListener that also listens to Source_CollectionChanged to remove the selection.

So what happens is that I set the new selection, and then it removes it. That causes weird behavior.

What would be the best work-around?

I'm pretty sure this was working before; at least better than it does currently. Anything changed in those libraries lately?

[FluentThemeSample.zip](https://github.com/AvaloniaUI/Avalonia/files/9229522/FluentThemeSample.zip)

The issue is the order the event handlers are being raised. The built-in handler needs to be raised first and mine second.

[Maybe learn from the Japanese: ](https://www.youtube.com/watch?v=IMtmoLoawBQ)
Woman: In Japan, men come first and women come second    
Austin Power: Or sometimes not at all!

**Comment by mysteryx93** [comment]
Solved it like this

```c#
private void Source_CollectionChanged(object? sender, NotifyCollectionChangedEventArgs e)
{
    // There is another event handler in the selection control that removes the current selection.
    // We must let it run first and select the item after.
    var pos = CurrentPosition;
    if (pos > -1)
    {
        Dispatcher.UIThread.Post(() => SetAndCoercePosition(pos), DispatcherPriority.Render);
    }
}
```

**Reply by mysteryx93** [reply]
This however leads to the problem that if I call this from code, I will assume that the last item will be selected but it won't be yet.

I had unit-tests in place to test that and the test fails. Because it's being tested, it was definitely working before. I wonder when it broke.

**Reply by timunie** [reply]
the Avalonia repo has a Sandbox App. Maybe you can create your sample in there and test several commits between working version and not working one. I think every release has a branch to restore it. So you can iterate over them (If you are interested in). I won't blame you if not ;-)

**Reply by mysteryx93** [reply]
I ended up solving the problem by clearing the selection before setting a new one, and it works. I'll leave it at that.
